misc/metrics/src/bandwidth.rs

```diff
@@ -0,0 +1,385 @@
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let this = self.project();
+        this.inner.poll_close(cx)
+    }
+}
+
+/// Allows obtaining the average bandwidth of the streams.
+#[derive(Default, Debug)]
+pub struct BandwidthSinks {
+    inbound: AtomicU64,
+    outbound: AtomicU64,
+}
+
+/// Wraps around an [`AsyncRead`] + [`AsyncWrite`] and logs the bandwidth that goes through it.
+#[pin_project::pin_project]
+pub struct InstrumentedStream<SMInner> {
+    #[pin]
+    inner: SMInner,
+    sinks: Arc<BandwidthSinks>,
+}
+
+impl<SMInner: AsyncRead> AsyncRead for InstrumentedStream<SMInner> {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        let this = self.project();
+        let num_bytes = ready!(this.inner.poll_read(cx, buf))?;
+        this.sinks.inbound.fetch_add(
+            u64::try_from(num_bytes).unwrap_or(u64::max_value()),
+            Ordering::Relaxed,
+        );
+        Poll::Ready(Ok(num_bytes))
+    }
+
+    fn poll_read_vectored(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        bufs: &mut [IoSliceMut<'_>],
+    ) -> Poll<io::Result<usize>> {
+        let this = self.project();
+        let num_bytes = ready!(this.inner.poll_read_vectored(cx, bufs))?;
+        this.sinks.inbound.fetch_add(
+            u64::try_from(num_bytes).unwrap_or(u64::max_value()),
+            Ordering::Relaxed,
+        );
+        Poll::Ready(Ok(num_bytes))
+    }
+}
+
+impl<SMInner: AsyncWrite> AsyncWrite for InstrumentedStream<SMInner> {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        let this = self.project();
+        let num_bytes = ready!(this.inner.poll_write(cx, buf))?;
+        this.sinks.outbound.fetch_add(
+            u64::try_from(num_bytes).unwrap_or(u64::max_value()),
+            Ordering::Relaxed,
+        );
+        Poll::Ready(Ok(num_bytes))
+    }
+
+    fn poll_write_vectored(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        bufs: &[IoSlice<'_>],
+    ) -> Poll<io::Result<usize>> {
+        let this = self.project();
+        let num_bytes = ready!(this.inner.poll_write_vectored(cx, bufs))?;
+        this.sinks.outbound.fetch_add(
+            u64::try_from(num_bytes).unwrap_or(u64::max_value()),
+            Ordering::Relaxed,
+        );
+        Poll::Ready(Ok(num_bytes))
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        let this = self.project();
+        this.inner.poll_flush(cx)
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        let this = self.project();
+        this.inner.poll_close(cx)
+    }
+}
+
+#[derive(Debug)]
+pub struct SinksCollector(Arc<RwLock<HashMap<String, Arc<BandwidthSinks>>>>);
```

Comment:
Why the indirection with a collector? Couldn't we just increment the counter directly? The counter metric is also just an atomic counter, right? So the performance should be similar just with less indirection overall.

Reply:
> Why the indirection with a collector? Couldn't we just increment the counter directly? The counter metric is also just an atomic counter, right? So the performance should be similar just with less indirection overall.

I am in favor of using `prometheus_client::counter::Counter` directly. I initially used a `Collector` in order to still enable folks to use their own metric system, i.e. map `BandwidthSinks` to their metric exposition mechanism.

> using locks instead of the atomic counters

Note that the locks are not in the hot path. A stream has a reference to the `BandwidthSinks` directly, without the need to lock a lock. That said, let me explore using `prometheus_client::counter::Counter` directly.